cmd/crypto/vault.go

```diff
@@ -123,86 +121,69 @@ func NewVault(config VaultConfig) (KMS, error) {
 		client.SetNamespace(config.Namespace)
 	}
 	v := &vaultService{client: client, config: &config}
-
 	if err := v.authenticate(); err != nil {
 		return nil, err
 	}
+	v.renewToken()
 	return v, nil
 }
 
-// renewSecret tries to renew the given secret. It blocks
-// until it receives either the new secret or encounters an error.
-func (v *vaultService) renewSecret(secret *vault.Secret) (*vault.Secret, error) {
-	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
-		Secret: secret,
-	})
-	if err != nil {
-		logger.CriticalIf(context.Background(), fmt.Errorf("crypto: failed to create hashicorp vault renewer: %s", err))
-	}
-	go renewer.Renew()
-	defer renewer.Stop()
-
-	for {
-		select {
-		case err := <-renewer.DoneCh():
-			if err != nil {
-				return nil, err
+// renewToken starts a new go-routine which renews
+// the vault authentication token periodically and re-authenticates
+// if the token renewal fails
+func (v *vaultService) renewToken() {
+	retryDelay := 1 * time.Minute
+	go func() {
+		for {
+			if v.secret == nil {
+				if err := v.authenticate(); err != nil {
+					time.Sleep(retryDelay)
+					continue
+				}
 			}
-		case renew := <-renewer.RenewCh():
-			if renew.Secret == nil || renew.Secret.Auth == nil {
-				return nil, ErrKMSAuthLogin
+			s, err := v.client.Auth().Token().RenewSelf(int(v.secret.Auth.LeaseDuration))
+			if err != nil || s == nil {
+				v.secret = nil
+				time.Sleep(retryDelay)
+				continue
 			}
-			return renew.Secret, nil
+			if ok, err := s.TokenIsRenewable(); !ok || err != nil {
+				v.secret = nil
+				continue
+			}
+			v.secret = s
+			nextRenew := s.Auth.LeaseDuration / 2
```

Comment:
Is it guaranteed that the `LeaseDuration` in https://godoc.org/github.com/hashicorp/vault/api#SecretAuth will be in sec and not e.g. minutes?

Reply:
@aead, internally their renewer assumes the lease duration is in seconds - so it should be safe to assume the same

Reply:
Yes, its guaranteed @aead - that's the logic even renewAuth() code has. https://www.vaultproject.io/api/system/leases.html#renew-lease

Reply:
Okay